lib/ProductOpener/Nutriscore.pm

```diff
@@ -430,5 +481,343 @@ sub compute_nutriscore_grade ($nutrition_score, $is_beverage, $is_water) {
+			and ($nutriscore_data_ref->{is_beverage})
+			and (defined $points_thresholds_2023{$nutrient_threshold_id . "_beverages"}))
+		{
+			$nutrient_threshold_id .= "_beverages";
+		}
+
+		$nutriscore_data_ref->{$nutrient . "_points"} = 0;
+
+		foreach my $threshold (@{$points_thresholds_2023{$nutrient_threshold_id}}) {
+			# The saturated fat ratio table uses the greater or equal sign instead of greater
+			if (   (($nutrient eq "saturated_fat_ratio") and ($nutriscore_data_ref->{$nutrient} >= $threshold))
+				or (($nutrient ne "saturated_fat_ratio") and ($nutriscore_data_ref->{$nutrient} > $threshold)))
+			{
+				$nutriscore_data_ref->{$nutrient . "_points"}++;
+			}
+		}
+	}
+
+	# For red meat products, the number of maximum protein points is set at 2 points
+	# Red meat products qualifying for this specific rule are products from beef, veal, swine and lamb,
+	# though they include also game/venison, horse, donkey, goat, camel and kangaroo.
+
+	if (($nutriscore_data_ref->{is_read_meat_products}) and ($nutriscore_data_ref->{proteins_points} > 2)) {
+		$nutriscore_data_ref->{proteins_points} = 2;
+	}
+
+	# Beverages with non-nutritive sweeteners have 4 extra negative points
+	if ($nutriscore_data_ref->{is_beverage}) {
+		if ($nutriscore_data_ref->{has_sweeteners}) {
+			$nutriscore_data_ref->{"sweeteners_points"} = 4;
+		}
+		else {
+			$nutriscore_data_ref->{"sweeteners_points"} = 0;
+		}
+	}
+
+	# Negative points
+
+	$nutriscore_data_ref->{negative_points} = 0;
+	foreach my $nutrient ($energy, "sugars", $saturated_fat, "salt", "sweeteners") {
+		$nutriscore_data_ref->{negative_points} += ($nutriscore_data_ref->{$nutrient . "_points"} || 0);
+	}
+
+	# Positive points
+
+	$nutriscore_data_ref->{positive_points} = 0;
+
+	my @positive_nutrients = qw(fruits_vegetables_legumes fiber);
+
+	# Positive points for proteins are counted in the following 3 cases:
+	# - the product is a beverage
+	# - the product is not in the fats, oils, nuts and seeds category
+	# and the negative points are less than 11 or the product is a cheese
+	# - the product is in the fats, oils, nuts and seeds category
+	# and the negative points are less than 7
+
+	if (
+		($nutriscore_data_ref->{is_beverage})
+		or (    (not $nutriscore_data_ref->{is_fat_oil_nuts_seeds})
+			and (($nutriscore_data_ref->{negative_points} < 11) or ($nutriscore_data_ref->{is_cheese})))
+		or ($nutriscore_data_ref->{negative_points} < 7)
+		)
+	{
+		push @positive_nutrients, "proteins";
+	}
+
+	foreach my $nutrient (@positive_nutrients) {
+		$nutriscore_data_ref->{positive_points} += ($nutriscore_data_ref->{$nutrient . "_points"} || 0);
+	}
+
+	my $score = $nutriscore_data_ref->{negative_points} - $nutriscore_data_ref->{positive_points};
+
+	return $score;
+}
+
+sub compute_nutriscore_grade_2023 ($nutrition_score, $is_beverage, $is_water) {
+
+	my $grade = "";
+
+	if (not defined $nutrition_score) {
+		return '';
+	}
+
+	if ($is_beverage) {
+
+		if ($is_water) {
+			$grade = 'a';
```

Comment:
this could be a nice place to include an explanatory comment in the structure

Reply:
The structure already has the structured data with the is_water field. The comment will need to be localized, it will be loaded in the knowledge panel template.